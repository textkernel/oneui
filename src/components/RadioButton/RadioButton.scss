@import '../../themes/oneui/oneui';

.RadioButton {

    &__box {
        background-color: var(--color-background);
        border: 1px solid var(--color-neutral-25);
        border-radius: 50%;
        box-sizing: content-box;
        display: inline-block;
        position: relative;
        top: -1px;
        width: 12px;
        height: 12px;
        transform: scale(1);
        transition: all var(--transition-duration) ease;
        vertical-align: middle;

        &:before {
            content: "";
            width: 100%;
            height: 100%;
            background: var(--color-primary);
            display: block;
            transform: scale(0);
            opacity: 1;
            border-radius: 50%;
        }

        &:focus,
        &:hover {
            border: 1px solid var(--color-primary);
            stroke: var(--color-highlight-background);
            background-color: var(--color-highlight-background);
        }
    }

    &__label {
        margin: auto;
        user-select: none;
        cursor: pointer;
        display: block;

        &:focus,
        &:hover {
            .RadioButton__box {
                border-color: var(--color-primary);
            }
        }
    }

    &__svg {
        position: absolute;
        top: 0;
        left: 0;
        fill: var(--color-background);
    }

    &__input {
        display: none;

        &:checked {
            + .RadioButton__label {
                .RadioButton__box {
                    background: var(--color-primary);
                    border-color: var(--color-primary);

                    .RadioButton__svg {
                        stroke-dashoffset: 0;
                    }
                }
            }
        }
    }

    &__text {
        padding: var(--spacing-normal) 0 var(--spacing-normal) var(--spacing-2x);
    }

    &--disabled {
        &:hover {
            .RadioButton__label {
                .RadioButton__box {
                    border-color: var(--color-neutral);
                }
            }
        }

        .RadioButton__input{
            + .RadioButton__label {
                cursor: not-allowed;
            }

            &:checked {
                + .RadioButton__label {
                    .RadioButton__box {
                        border-color: var(--color-neutral);
                        background: var(--color-neutral);
                    }
                }
            }
        }
    }
}
