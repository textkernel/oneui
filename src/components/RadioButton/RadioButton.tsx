import React, { forwardRef } from 'react';
import { bem } from '../../utils';
import { Text } from '../Text';
import styles from './RadioButton.scss';
import { CHECKBOX_VIEWBOX } from '../../constants';
import { SingleReactNode } from '../../customTypes/types';

export interface Props extends React.HTMLProps<HTMLInputElement> {
    /** A unique id to reference this radio button */
    id: string;
    /** The name of the group this radio button belongs to */
    name?: string;
    /** If the radio button should be disabled */
    disabled?: boolean;
    /** The label for the radio button */
    children?: SingleReactNode;
}

const { block, elem } = bem('RadioButton', styles);

export const RadioButton = forwardRef<HTMLElement, Props>(
    ({ id, children, disabled = false, name, ...rest }, ref) => {
        return (
            <div {...block({ ...rest, disabled })}>
                <input
                    {...rest}
                    {...elem('input')}
                    ref={ref}
                    type="radio"
                    id={id}
                    name={name}
                    disabled={disabled}
                />
                <label {...elem('label')} htmlFor={id}>
                    <span {...elem('box')}>
                        <svg {...elem('svg')} width="12px" height="10px" viewBox={CHECKBOX_VIEWBOX}>
                            <circle cx="5" cy="6" r="3" />
                        </svg>
                    </span>
                    {!!children && (
                        <Text {...elem('text')} inline context={disabled ? 'muted' : 'default'}>
                            {children}
                        </Text>
                    )}
                </label>
            </div>
        );
    }
);

RadioButton.displayName = 'RadioButton';
