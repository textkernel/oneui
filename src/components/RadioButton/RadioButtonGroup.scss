.RadioButtonGroup__option + .RadioButtonGroup__option {
    margin-top: var(--space-75);
}
