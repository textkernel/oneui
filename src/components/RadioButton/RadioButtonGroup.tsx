import React, { forwardRef } from 'react';
import { bem } from '../../utils';
import styles from './RadioButtonGroup.scss';

export interface Props {
    /** The name of the group this radio button belongs to */
    name: string;
    /** common onChange handler that will be passed to all children */
    onChange?: React.ChangeEventHandler<HTMLInputElement> | undefined;
    /** The radio buttons */
    children: null | undefined | React.ReactElement[];
}

const { block, elem } = bem('RadioButtonGroup', styles);

export const RadioButtonGroup = forwardRef<HTMLElement, Props>(
    ({ children, name, onChange, ...rest }, ref) => {
        const childProps = onChange === undefined ? { name } : { name, onChange };

        return (
            <div ref={ref} {...rest} {...block({ ...rest })}>
                {React.Children.map(children, (child) => {
                    if (child) {
                        return React.cloneElement(child, {
                            ...elem('option', {
                                elemClassName: child.props.className,
                            }),
                            ...childProps,
                        });
                    }
                    return null;
                })}
            </div>
        );
    }
);

RadioButtonGroup.displayName = 'RadioButtonGroup';
