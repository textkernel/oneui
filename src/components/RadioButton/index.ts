export { RadioButton, Props as RadioButtonProps } from './RadioButton';
export { RadioButtonGroup, Props as RadioButtonGroupProps } from './RadioButtonGroup';
