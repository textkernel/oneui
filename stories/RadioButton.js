import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { RadioButton, RadioButtonGroup } from '@textkernel/oneui';

storiesOf('Molecules/RadioButton', module)
    .addDecorator(withKnobs)
    .add('RadioButton', () => (
        <RadioButtonGroup
            name="my-group"
            onChange={(e) =>
                console.log('onChange was called with event.target.value: ', e.target.value)
            }
        >
            <RadioButton disabled={boolean('Disabled 1', false)} value="option1" id="radio-1">
                {text('RadioButton 1 label', 'Option 1')}
            </RadioButton>
            <RadioButton disabled={boolean('Disabled 2', false)} value="option2" id="radio-2">
                {text('RadioButton 2 label', 'Option 2')}
            </RadioButton>
        </RadioButtonGroup>
    ));
