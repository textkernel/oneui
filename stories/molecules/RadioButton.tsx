import * as React from 'react';
import { RadioButton, RadioButtonGroup } from '@textkernel/oneui';

export default {
    title: 'Molecules/RadioButton',
    component: RadioButton,
    subcomponents: { RadioButtonGroup },
};

export const _RadioButton = (args) => (
    <RadioButtonGroup name="my-group">
        <RadioButton {...args} />
        <RadioButton id="radio-2" value="option2">
            Option 2
        </RadioButton>
    </RadioButtonGroup>
);

_RadioButton.args = {
    value: 'option1',
    id: 'radio-1',
    children: 'Option 1',
};
