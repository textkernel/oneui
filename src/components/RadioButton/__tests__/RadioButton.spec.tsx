import React from 'react';
import toJson from 'enzyme-to-json';
import { RadioButton } from '../RadioButton';

describe('<RadioButton> that renders a radio button', () => {
    it('should render default radio button correctly', () => {
        const wrapper = mount(<RadioButton id="c1" />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render radio button with props and children correctly', () => {
        const wrapper = mount(
            <RadioButton id="c1" name="group_name">
                Choose me
            </RadioButton>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should call onChange function when clicked', () => {
        const onChange = jest.fn();
        const wrapper = mount(
            <RadioButton id="c2" onChange={onChange}>
                Choose me
            </RadioButton>
        );
        wrapper.find('input').simulate('change');
        expect(onChange).toHaveBeenCalledTimes(1);
    });
    it('should rendered disabled radio button correctly', () => {
        const wrapper = mount(
            <RadioButton id="c3" disabled>
                Useless radio button
            </RadioButton>
        );
        expect(wrapper.find('input[disabled]')).toHaveLength(1);
        expect(wrapper.find('.Text--context_muted')).toHaveLength(1);
    });
});
