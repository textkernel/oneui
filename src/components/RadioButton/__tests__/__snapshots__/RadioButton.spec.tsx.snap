// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton> that renders a radio button should render default radio button correctly 1`] = `
<RadioButton
  id="c1"
>
  <div>
    <input
      className="RadioButton__input"
      disabled={false}
      id="c1"
      type="radio"
    />
    <label
      className="RadioButton__label"
      htmlFor="c1"
    >
      <span
        className="RadioButton__box"
      >
        <svg
          className="RadioButton__svg"
          height="10px"
          viewBox={
            Array [
              0,
              0,
              10,
              10,
            ]
          }
          width="12px"
        >
          <circle
            cx="5"
            cy="6"
            r="3"
          />
        </svg>
      </span>
    </label>
  </div>
</RadioButton>
`;

exports[`<RadioButton> that renders a radio button should render radio button with props and children correctly 1`] = `
<RadioButton
  id="c1"
  name="group_name"
>
  <div>
    <input
      className="RadioButton__input"
      disabled={false}
      id="c1"
      name="group_name"
      type="radio"
    />
    <label
      className="RadioButton__label"
      htmlFor="c1"
    >
      <span
        className="RadioButton__box"
      >
        <svg
          className="RadioButton__svg"
          height="10px"
          viewBox={
            Array [
              0,
              0,
              10,
              10,
            ]
          }
          width="12px"
        >
          <circle
            cx="5"
            cy="6"
            r="3"
          />
        </svg>
      </span>
      <Text
        className="RadioButton__text"
        context="default"
        inline={true}
      >
        <span
          className="RadioButton__text"
        >
          Choose me
        </span>
      </Text>
    </label>
  </div>
</RadioButton>
`;

exports[`<RadioButton> that renders a radio button should rendered disabled radio button correctly 1`] = `
<RadioButton
  disabled={true}
  id="c3"
>
  <div
    className="RadioButton--disabled"
  >
    <input
      className="RadioButton__input"
      disabled={true}
      id="c3"
      type="radio"
    />
    <label
      className="RadioButton__label"
      htmlFor="c3"
    >
      <span
        className="RadioButton__box"
      >
        <svg
          className="RadioButton__svg"
          height="10px"
          viewBox={
            Array [
              0,
              0,
              10,
              10,
            ]
          }
          width="12px"
        >
          <circle
            cx="5"
            cy="6"
            r="3"
          />
        </svg>
      </span>
      <Text
        className="RadioButton__text"
        context="muted"
        inline={true}
      >
        <span
          className="RadioButton__text"
        >
          Useless radio button
        </span>
      </Text>
    </label>
  </div>
</RadioButton>
`;
